### PR TITLE
Add evaluation metrics script

### PR DESCRIPTION
**Pull request by Soorya19Pradeep**
The evaluation metrics is computed using two scripts here. 

- `curator_script.py` creates cellpose segmented masks for target channels on the original fluorescence images which are saved as pngs and predictions, which are saved as a new channel on the zarr. 
- The user can correct the masks generated using cellpose and original fluorescence images and overwrite the masks through cellpose gui.
- `metrics_script.py` reads in both predicted image mask and user corrected ground truth mask and computes the user-listed metrics values.

I have transferred the changes created by @JohannaRahm on metrics calculation functions made in sandbox on this branch.

### CODE REVIEW

File: micro_dl/cli/curator_script.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
+# %% script to generate your ground truth directory for microDL prediction evaluation
+# After inference, the predictions generated are stored as zarr store.
+# Evaluation metrics can be computed by comparison of prediction to human proof read ground truth.
+#
+import numpy as np
+import os
+from PIL import Image
+import imagio as iio
```
Comment (mattersoflight) [inline micro_dl/cli/curator_script.py:8]: imageio needs to be in `lit-monai` environment

File: micro_dl/cli/curator_script.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+import iohub.ngff as ngff
+import argparse
+
+import micro_dl.inference.evaluation_metrics as metrics
+import micro_dl.utils.aux_utils as aux_utils
+from micro_dl.utils.mp_utils import add_channel
+from waveorder.focus import focus_from_transverse_band
```
Comment (mattersoflight) [inline micro_dl/cli/curator_script.py:15]: waveorder needs to be in `lit-monai` environment

**Review by mattersoflight — CHANGES_REQUESTED** [review]
The overall structure looks good! This PR seems ready for review. 

@Soorya19Pradeep please point to the config and data suitable for testing the branch. 
@ziw-liu please update the `lit-moai` environment with dependencies needed by the curation/evaluation CLI.

File: micro_dl/inference/evaluation_metrics.py [hunk 3]
```diff
@@ -4,8 +4,13 @@
 import pandas as pd
 from skimage.metrics import structural_similarity as ssim
 import sklearn.metrics
+from skimage.morphology import disk, dilation, erosion
+from skimage.measure import label, regionprops
 from scipy.stats import pearsonr
 
+from cellpose import models, utils, io
+from lapsolver import solve_dense
+
```
[the remark this answers is not in the document]
Comment (ziw-liu) [reply micro_dl/inference/evaluation_metrics.py:13]: @Soorya19Pradeep Can you provide versions that you depend on for these libraries? I know that sklearn and cellpose both have had some breaking API changes over the recent years.

File: micro_dl/cli/curator_script.py (hunk 1 above)
Comment (ziw-liu) [reply micro_dl/cli/curator_script.py:8]: Assuming that `imagio` here means `imageio`, it is already in the env (as required by skimage).

File: micro_dl/cli/curator_script.py (hunk 2 above)
Comment (ziw-liu) [reply micro_dl/cli/curator_script.py:15]: Done:
```text
$ pip show waveorder
Name: waveorder
Version: 1.0.0rc1
```

File: micro_dl/cli/curator_script.py (hunk 1 above)
Comment (ziw-liu) [reply micro_dl/cli/curator_script.py:8]:
```suggestion
import imageio as iio
```

**Comment by Soorya19Pradeep** [comment]
> The overall structure looks good! This PR seems ready for review.
> 
> @Soorya19Pradeep please point to the config and data suitable for testing the branch. @ziw-liu please update the `lit-moai` environment with dependencies needed by the curation/evaluation CLI.

I have not been able to extensively test this code using the actual data. I developed it around a separately generated mock config file and a copy of pred.zarr that was generated by @mattersoflight testing the most recent changes in training pipeline. The config I created is here:
`/local/scratch/groups/cmanalysis.grp/microDL_SP/config/eval_test/torch_config_Phase_deconNucl_EvalMetricsTest.yml`

The copy of pred.zarr I used is here: 
`/hpc/projects/CompMicro/projects/infected_cell_imaging/VirtualStaining/VirtualStain_NuclMem_HEK_2022/model_nucl25d/training_model_2023_03_17_mock_test/inference_results_2023_03_17_mock_test/preds.zarr`

File: micro_dl/inference/evaluation_metrics.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ziw-liu) [reply micro_dl/inference/evaluation_metrics.py:13]: Installed the versions @Soorya19Pradeep provided, excpet for lapsolver, for which I installed `1.1.0` instead of `1.1.0=dev_0` since the later seems to be a typo.

File: micro_dl/inference/evaluation_metrics.py [hunk 4]
```diff
@@ -165,6 +319,18 @@ def binarize_array(im):
...
+def cpmask_array(im, model_name):
+    """mask preparation using cellpose
+
+    :input model: cellpose model used for segmentation
+    :input im: predicted image to be segmented
+    :return im_mask: cellpose segmented mask
+    """
+    model = models.CellposeModel(model_type=model_name)
+    channels = [0, 0]
+    masks = model.eval(im, channels=channels, diameter=None)[0]
```
Comment (ziw-liu) [inline micro_dl/inference/evaluation_metrics.py:331]: Nothing is returned here.

Comment (Soorya19Pradeep) [reply micro_dl/inference/evaluation_metrics.py:331]: Thanks for pointing out! I have fixed it.

File: micro_dl/cli/curator_script.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+                        cp_mask,
+                    )  # save binary mask as numpy or png
+
+    # segment prediction and add mask as channel to pred_dir
+    pred_plate = ngff.open_ome_zarr(store_path=pred_dir, mode="r+")
+    # im_pred = pred_plate.data
+    chan_names = pred_plate.channel_names
+
+    for channel_name in chan_names:
+        for i, (position, pos_data) in enumerate(pred_plate.positions()):
```
Comment (ziw-liu) [inline micro_dl/cli/curator_script.py:146]: No need for `enumerate()` if the index is not used.

Also, it will be more efficient iterate over positions in the outer loop, since opening positions are more expensive.

Comment (Soorya19Pradeep) [reply micro_dl/cli/curator_script.py]: I have removed the unused enumeration in two places

File: micro_dl/cli/metrics_script.py [hunk 6]
```diff
@@ -1,263 +1,138 @@
...
+
+        if pos in PosList:
+            raw_data = pos_data.data
+            target_data = raw_data[0, chan_names.index(metric_chan_mask), ...]
+
+            gt_mask_save_name = (
+                "^" + metric_channel + "_p" + str(format(pos, "03d")) + "_z"
+            )
+            z_slice_no = int(
+                glob.glob(ground_truth_dir + "/" + gt_mask_save_name + "*_cp_mask.png")
```
Comment (ziw-liu) [inline micro_dl/cli/metrics_script.py:110]: This will not work on Windows. `os.path.join` is more robust to file system variations.

**Review by ziw-liu — APPROVED** [review]
